prow/tide/gerrit.go

```diff
@@ -0,0 +1,320 @@
+		return nil, fmt.Errorf("Cannot list prowjob with selector %v", selector)
+	}
+
+	// keep track of latest prowjobs only
+	latestPjs := make(map[string]*prowapi.ProwJob)
+	for _, pj := range pjs.Items {
+		pj := pj
+		exist, ok := latestPjs[pj.Spec.Context]
+		if ok && exist.CreationTimestamp.After(pj.CreationTimestamp.Time) {
+			continue
+		}
+		latestPjs[pj.Spec.Context] = &pj
+	}
+
+	for _, pj := range latestPjs {
+		res = append(res, Context{
+			Context:     githubql.String(pj.Spec.Context),
+			Description: githubql.String(config.ContextDescriptionWithBaseSha(pj.Status.Description, pj.Spec.Refs.BaseSHA)),
+			State:       githubql.StatusState(pj.Status.State),
+		})
+	}
+
+	return res, nil
+}
+
+func (p *GerritProvider) mergePRs(sp subpool, prs []CodeReviewCommon, dontUpdateStatus *threadSafePRSet) error {
+	p.logger.Info("The merge function hasn't been implemented yet, just logging for now.")
+	return nil
+}
+
+// GetTideContextPolicy gets context policy defined by users + requirements from
+// prow jobs.
+//
+// These are all out of scope for Gerrit, as Gerrit PRs queried are already
+// submittable and mergeable. So return an empty policy option should do.
+func (p *GerritProvider) GetTideContextPolicy(gitClient git.ClientFactory, org, repo, branch string, baseSHAGetter config.RefGetter, crc *CodeReviewCommon) (contextChecker, error) {
+	pr := crc.Gerrit
+	if pr == nil {
+		return nil, errors.New("error not supposed to happen, crc.Gerrit cannot be nil for GerritProvider")
+	}
+
+	required := sets.NewString()
+	requiredIfPresent := sets.NewString()
+	optional := sets.NewString()
+
+	headSHAGetter := func() (string, error) {
+		return crc.HeadRefOID, nil
+	}
+	presubmits, err := p.cfg().GetPresubmits(gitClient, org+"/"+repo, baseSHAGetter, headSHAGetter)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get presubmits: %w", err)
+	}
+
+	requireLabels := sets.NewString()
+	for l, info := range pr.Labels {
+		if !info.Optional {
+			requireLabels.Insert(l)
+		}
+	}
+
+	// automatically generate required and optional entries for Prow Jobs
+	for _, pj := range presubmits {
+		if !pj.CouldRun(branch) {
+			continue
+		}
+
+		var isJobRequired bool
+		if val, ok := pj.Labels[kube.GerritReportLabel]; ok && requireLabels.Has(val) {
+			isJobRequired = true
+		}
+
+		if isJobRequired {
+			if pj.TriggersConditionally() {
+				// jobs that trigger conditionally cannot be
+				// required as their status may not exist on PRs
+				requiredIfPresent.Insert(pj.Context)
+			} else {
+				// jobs that produce required contexts and will
+				// always run should be required at all times
+				required.Insert(pj.Context)
+			}
+		} else {
+			optional.Insert(pj.Context)
+		}
+	}
+
+	t := &config.TideContextPolicy{
+		RequiredContexts:          required.List(),
+		RequiredIfPresentContexts: requiredIfPresent.List(),
+		OptionalContexts:          optional.List(),
+	}
+	if err := t.Validate(); err != nil {
+		return t, err
```

Comment:
We should avoid logging this at the error level since it is a user configuration problem. Ideally we'd either expose this to users or better yet, catch this before it happens with presubmit validation.

Reply:
this is from https://github.com/kubernetes/test-infra/blob/d24f409b8fa140e327afc12c08f039379221e516/prow/config/tide.go#L703, which is basically the implementation of this function for GitHub, seems like it's been working well, so we might be fine here?

Reply:
Hmm, well hopefully it is impossible to hit this for Gerrit as well. If we reach it somehow we can make changes then.

Reply:
ack